doc/usage.rst

```diff
@@ -98,6 +112,8 @@ Complete configuration for rules can be supplied using a ``json`` formatted stri
 
 The ``--dry-run`` flag will run the fixer without making changes to your files (implicitly set when you use ``check`` command).
 
+The ``--sequential`` flag will enforce sequential analysis even if parallel config is provided.
```

Comment:
I would rather not want to change the current approach. I believe parallel config should be provided explicitly in the config, because it's not only about CPUs, but also batch size and timeout. Yeah, there are default values for these params, but if we wanted to add CLI options for that, IMHO we would need to support all of them.. And I'm kind of against it 🤔.

Reply:
I do not understand a case when timeout should be configured differently than default.

I expect that auto-discovery of CPU is good enough, but I can imagine someone in need to configure it explicitly.
But for that, this is not config per project, but config per environment - I either want `auto` everywhere, or `auto` on my friend's dev machine, fixed `4` on my CI, and crippled hardcoded `1` on my machine when most of CPU I want to reserve for some other special process (whatever it could be, dunno).


What I am trying to say is that I'm kind-of against having per-project, committed to repo config of 4 or 5 or 7 or 12 CPUs. It shall be `auto` or `1`.
And if someone needs it dynamic, it's runtime and not project config.

Reply:
> I do not understand a case when timeout should be configured differently than default.

Timeout is actually per each main process' request (after establishing connection, requesting new batch etc), not per whole process. So it's configurable because if you prefer smaller batches, you can set smaller timeout, but if you bump batch size, then most probably you would need bigger timeout too.

> What I am trying to say is that I'm kind-of against having per-project, committed to repo config of 4 or 5 or 7 or 12 CPUs. It shall be auto or 1.
> And if someone needs it dynamic, it's runtime and not project config.

I don't agree with this actually. In GetResponse we have in `rector.php`:

```php
return static function (RectorConfig $rectorConfig): void {
    $maxNumberOfProcesses = 16;

    if (isset($_ENV['CI'])) {
        // Limit parallel analysis when running on Gitlab CI
        $maxNumberOfProcesses = 4;
    } elseif (isset($_ENV['RECTOR_PARALLEL_MAX_NUMBER_OF_PROCESSES']) && is_numeric($_ENV['RECTOR_PARALLEL_MAX_NUMBER_OF_PROCESSES'])) {
        // allow configurable number of processes for local usage
        $maxNumberOfProcesses = (int)$_ENV['RECTOR_PARALLEL_MAX_NUMBER_OF_PROCESSES'];
    }

    $rectorConfig->parallel(
        600,
        $maxNumberOfProcesses
    );

    // ...
};
```

So it's the same config for local development and for the CI. Since it's a PHP config, just like in Fixer, you can do whatever you want to allow dynamic config. Putting everything to CLI input is not good IMHO. None of PHPStan, Rector or ECS has parallel configuration through CLI options. ParaTest has it, but parallelisation is its core concept.

Also, from my personal experience - I run Fixer both on locally installed PHP and through Docker. CPU auto-detection works properly on both runtimes (as you can see [here](#auto-detection)). But if I wanted to limit CPUs used by Fixer, I really prefer to do it in a local config rather than pass it all the time via CLI options. Especially because you can have dist config and then do in `.php-cs-fixer.php`:

```php
ini_set('memory_limit', '-1');

$config = require(__DIR__ . '/.php-cs-fixer.dist.php');
$config->setParallelConfig(new ParallelConfig(2, 50, 300));

return $config;
```

This is bare minimum config override, but of course you can do more than that. It's really better approach than adding `--cpu=X` all the time.

Reply:
@mfn thanks for the input! As you can see in my [previous comment](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7777#discussion_r1526983794) you can achieve dynamically changed CPUs via CLI by defining config file with support for this. In case of the GR's Rector config you could just run Rector with `RECTOR_PARALLEL_MAX_NUMBER_OF_PROCESSES=4 vendor/bin/rector [...]` and the config would set it based on the one-time env variable. Since Fixer's config also is PHP-based, you can implement similar approach in your projects' dist config. That's why I'm pretty sure we don't need to support this out-of-the-box, and especially not in this PR ☺️.